Q1 Load this data in sql and in pandas with a relation in SQL.

In [8]:
import pandas as pd
import pymysql
pd.set_option('display.max_columns', None)

In [6]:
store = pd.read_excel('Superstore_USA.xlsx')

In [10]:
store.to_csv('USA_Superstore.csv', index = False)

In [11]:
store.head()

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88           0.50            2   
1   20847           High      0.01        2.84           0.93            3   
2   23086  Not Specified      0.03        6.68           6.15            3   
3   23087  Not Specified      0.01        5.68           3.60            3   
4   23088  Not Specified      0.00      205.99           2.50            3   

     Customer Name    Ship Mode Customer Segment Product Category  \
0  Janice Fletcher  Regular Air        Corporate  Office Supplies   
1    Bonnie Potter  Express Air        Corporate  Office Supplies   
2    Bonnie Potter  Express Air        Corporate  Office Supplies   
3    Bonnie Potter  Regular Air        Corporate  Office Supplies   
4    Bonnie Potter  Express Air        Corporate       Technology   

            Product Sub-Category Product Container  \
0                         Labels         Small Box   
1            Pens & Art Supplies          Wrap Bag   
2                          Paper         Small Box   
3  Scissors, Rulers and Trimmers        Small Pack   
4   Telephones and Communication         Small Box   

                                     Product Name  Product Base Margin  \
0                                        Avery 49                 0.36   
1  SANFORD Liquid Accent™ Tank-Style Highlighters                 0.54   
2                                      Xerox 1968                 0.37   
3        Acme® Preferred Stainless Steel Scissors                 0.56   
4                                             V70                 0.59   

    Region State or Province       City  Postal Code Order Date  Ship Date  \
0  Central          Illinois    Addison        60101 2012-05-28 2012-05-30   
1     West        Washington  Anacortes        98221 2010-07-07 2010-07-08   
2     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
3     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
4     West        Washington  Anacortes        98221 2011-07-27 2011-07-27   

     Profit  Quantity ordered new    Sales  Order ID  
0    1.3200                     2     5.90     88525  
1    4.5600                     4    13.01     88522  
2  -47.6400                     7    49.92     88523  
3  -30.5100                     7    41.64     88523  
4  998.2023                     8  1446.67     88523

Q2 While loading this data you dont have to create a table manually you can use any automated approach to create a table and load data in bulk in table.

In [12]:
dbcon = pymysql.connect(host = "localhost", user = "root", password  , database = "ASSIGNMENT")

In [13]:
dbcon

In [14]:
data = pd.read_sql_query("""SELECT * FROM usa_superstore""", dbcon)

In [15]:
data.head()

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88           0.50            2   
1   20847           High      0.01        2.84           0.93            3   
2   23086  Not Specified      0.03        6.68           6.15            3   
3   23087  Not Specified      0.01        5.68           3.60            3   
4   23088  Not Specified      0.00      205.99           2.50            3   

     Customer Name    Ship Mode Customer Segment Product Category  \
0  Janice Fletcher  Regular Air        Corporate  Office Supplies   
1    Bonnie Potter  Express Air        Corporate  Office Supplies   
2    Bonnie Potter  Express Air        Corporate  Office Supplies   
3    Bonnie Potter  Regular Air        Corporate  Office Supplies   
4    Bonnie Potter  Express Air        Corporate       Technology   

            Product Sub-Category Product Container  \
0                         Labels         Small Box   
1            Pens & Art Supplies          Wrap Bag   
2                          Paper         Small Box   
3  Scissors, Rulers and Trimmers        Small Pack   
4   Telephones and Communication         Small Box   

                                    Product Name  Product Base Margin  \
0                                       Avery 49                 0.36   
1  SANFORD Liquid Accent Tank-Style Highlighters                 0.54   
2                                     Xerox 1968                 0.37   
3        Acme Preferred Stainless Steel Scissors                 0.56   
4                                            V70                 0.59   

    Region State or Province       City  Postal Code  Order Date   Ship Date  \
0  Central          Illinois    Addison        60101  2012-05-28  2012-05-30   
1     West        Washington  Anacortes        98221  2010-07-07  2010-07-08   
2     West        Washington  Anacortes        98221  2011-07-27  2011-07-28   
3     West        Washington  Anacortes        98221  2011-07-27  2011-07-28   
4     West        Washington  Anacortes        98221  2011-07-27  2011-07-27   

     Profit  Quantity ordered new    Sales  Order ID  
0    1.3200                     2     5.90     88525  
1    4.5600                     4    13.01     88522  
2  -47.6400                     7    49.92     88523  
3  -30.5100                     7    41.64     88523  
4  998.2023                     8  1446.67     88523

In [26]:
store.shape

(9426, 24)

In [19]:
store.describe()

Row ID     Discount   Unit Price  Shipping Cost  Customer ID  \
count   9426.000000  9426.000000  9426.000000    9426.000000  9426.000000   
mean   20241.015277     0.049628    88.303686      12.795142  1738.422236   
std     6101.890965     0.031798   281.540982      17.181203   979.167197   
min        2.000000     0.000000     0.990000       0.490000     2.000000   
25%    19330.250000     0.020000     6.480000       3.192500   898.000000   
50%    21686.500000     0.050000    20.990000       6.050000  1750.000000   
75%    24042.750000     0.080000    85.990000      13.990000  2578.750000   
max    26399.000000     0.250000  6783.020000     164.730000  3403.000000   

       Product Base Margin   Postal Code        Profit  Quantity ordered new  \
count          9354.000000   9426.000000   9426.000000           9426.000000   
mean              0.512189  52446.327286    139.236410             13.798430   
std               0.135229  29374.597802    998.486483             15.107688   
min               0.350000   1001.000000 -16476.838000              1.000000   
25%               0.380000  29406.000000    -74.017375              5.000000   
50%               0.520000  52302.000000      2.567600             10.000000   
75%               0.590000  78516.000000    140.243850             17.000000   
max               0.850000  99362.000000  16332.414000            170.000000   

               Sales      Order ID  
count    9426.000000   9426.000000  
mean      949.706272  82318.489073  
std      2598.019818  19149.448857  
min         1.320000      6.000000  
25%        61.282500  86737.250000  
50%       203.455000  88344.500000  
75%       776.402500  89987.750000  
max    100119.160000  91591.000000

In [17]:
store.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9426 entries, 0 to 9425
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Row ID                9426 non-null   int64         
 1   Order Priority        9426 non-null   object        
 2   Discount              9426 non-null   float64       
 3   Unit Price            9426 non-null   float64       
 4   Shipping Cost         9426 non-null   float64       
 5   Customer ID           9426 non-null   int64         
 6   Customer Name         9426 non-null   object        
 7   Ship Mode             9426 non-null   object        
 8   Customer Segment      9426 non-null   object        
 9   Product Category      9426 non-null   object        
 10  Product Sub-Category  9426 non-null   object        
 11  Product Container     9426 non-null   object        
 12  Product Name          9426 non-null   object        
 13  Product Base Margi

In [18]:
store.isnull().sum()

Row ID                   0
Order Priority           0
Discount                 0
Unit Price               0
Shipping Cost            0
Customer ID              0
Customer Name            0
Ship Mode                0
Customer Segment         0
Product Category         0
Product Sub-Category     0
Product Container        0
Product Name             0
Product Base Margin     72
Region                   0
State or Province        0
City                     0
Postal Code              0
Order Date               0
Ship Date                0
Profit                   0
Quantity ordered new     0
Sales                    0
Order ID                 0
dtype: int64

In [30]:
store['Product Name'].value_counts()

Global High-Back Leather Tilter, Burgundy                            27
Bevis 36 x 72 Conference Tables                                      26
BoxOffice By Design Rectangular and Half-Moon Meeting Room Tables    24
Master Giant Foot® Doorstop, Safety Yellow                           24
Wilson Jones Hanging View Binder, White, 1"                          23
                                                                     ..
Fellowes High-Stak® Drawer Files                                      1
Computer Printout Index Tabs                                          1
Hewlett Packard 610 Color Digital Copier / Printer                    1
Global Comet™ Stacking Armless Chair                                  1
Alliance Rubber Bands                                                 1
Name: Product Name, Length: 1263, dtype: int64

Q3 Find out how much return we have recieved with a product name.

In [31]:
store.groupby('Product Name')['Profit'].sum()

Product Name
"While you Were Out" Message Book, One Form per Page     708.820800
#10 Self-Seal White Envelopes                            223.197520
#10 White Business Envelopes,4 1/8 x 9 1/2              1164.001460
#10- 4 1/8" x 9 1/2" Recycled Envelopes                 -888.757800
#10- 4 1/8" x 9 1/2" Security-Tint Envelopes           -1288.400800
                                                           ...     
i500plus                                                 808.244640
i600                                                    -525.934816
iDEN i550                                               2624.076000
iDEN i95                                                -371.274704
iDENi80s                                                -471.229440
Name: Profit, Length: 1263, dtype: float64

Q4 Try to join order and return data both in sql and pandas.

In [35]:
ordr = pd.DataFrame(store['Order ID'])

In [36]:
rtn = pd.DataFrame(store['Profit'])

In [37]:
ordr.join(rtn)

Order ID       Profit
0        88525     1.320000
1        88522     4.560000
2        88523   -47.640000
3        88523   -30.510000
4        88523   998.202300
...        ...          ...
9421     87532   137.860000
9422     87532   -39.070000
9423     87533 -1225.029097
9424     87530   349.485000
9425     87533    78.062400

[9426 rows x 2 columns]

In [38]:
data = pd.read_sql_query("""SELECT `Order ID`, Profit FROM usa_superstore""", dbcon)

In [39]:
data

Order ID     Profit
0       88525     1.3200
1       88522     4.5600
2       88523   -47.6400
3       88523   -30.5100
4       88523   998.2023
..        ...        ...
167     87414   -10.6582
168     87421  -218.9761
169     87948  1312.0380
170     87948  1451.9394
171     87951    95.3787

[172 rows x 2 columns]

Q5 Try to find out how many unique customers that we have.

In [43]:
store['Customer ID'].nunique()

2703

WE HAVE 2703 UNIQUE CUSTOMERS IN TOTAL

In [44]:
store['Customer Name'].nunique()

2703

Every customer has a unique name too.

Q6 Try to find out in how many regions we are selling a product and who is the manager for the respective region.

In [50]:
store['Region'].value_counts()

Central    2899
East       2289
West       2284
South      1954
Name: Region, dtype: int64

In [52]:
store.groupby('Region')['Customer Segment'].value_counts()

Region   Customer Segment
Central  Corporate           1196
         Home Office          749
         Consumer             481
         Small Business       473
East     Corporate            726
         Small Business       581
         Home Office          519
         Consumer             463
South    Corporate            692
         Home Office          503
         Consumer             433
         Small Business       326
West     Corporate            761
         Home Office          545
         Consumer             517
         Small Business       461
Name: Customer Segment, dtype: int64

Q7 Find out how many different shipment mode that we have and what is the percentage usability of all the shipment mode with respect to dataset.

In [53]:
store.head()

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88           0.50            2   
1   20847           High      0.01        2.84           0.93            3   
2   23086  Not Specified      0.03        6.68           6.15            3   
3   23087  Not Specified      0.01        5.68           3.60            3   
4   23088  Not Specified      0.00      205.99           2.50            3   

     Customer Name    Ship Mode Customer Segment Product Category  \
0  Janice Fletcher  Regular Air        Corporate  Office Supplies   
1    Bonnie Potter  Express Air        Corporate  Office Supplies   
2    Bonnie Potter  Express Air        Corporate  Office Supplies   
3    Bonnie Potter  Regular Air        Corporate  Office Supplies   
4    Bonnie Potter  Express Air        Corporate       Technology   

            Product Sub-Category Product Container  \
0                         Labels         Small Box   
1            Pens & Art Supplies          Wrap Bag   
2                          Paper         Small Box   
3  Scissors, Rulers and Trimmers        Small Pack   
4   Telephones and Communication         Small Box   

                                     Product Name  Product Base Margin  \
0                                        Avery 49                 0.36   
1  SANFORD Liquid Accent™ Tank-Style Highlighters                 0.54   
2                                      Xerox 1968                 0.37   
3        Acme® Preferred Stainless Steel Scissors                 0.56   
4                                             V70                 0.59   

    Region State or Province       City  Postal Code Order Date  Ship Date  \
0  Central          Illinois    Addison        60101 2012-05-28 2012-05-30   
1     West        Washington  Anacortes        98221 2010-07-07 2010-07-08   
2     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
3     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
4     West        Washington  Anacortes        98221 2011-07-27 2011-07-27   

     Profit  Quantity ordered new    Sales  Order ID  
0    1.3200                     2     5.90     88525  
1    4.5600                     4    13.01     88522  
2  -47.6400                     7    49.92     88523  
3  -30.5100                     7    41.64     88523  
4  998.2023                     8  1446.67     88523

In [55]:
df = pd.DataFrame(store['Ship Mode'].value_counts())

In [56]:
df

Ship Mode
Regular Air          7036
Delivery Truck       1283
Express Air          1107

In [58]:
df['percent'] = (df['Ship Mode'] / df['Ship Mode'].sum()) * 100

In [59]:
df

Ship Mode    percent
Regular Air          7036  74.644600
Delivery Truck       1283  13.611288
Express Air          1107  11.744112

Q8 Create a new coulmn and try to find the difference between order date and shipment date.

In [60]:
store['Order Date'] = pd.to_datetime(store['Order Date'])

In [61]:
store['Ship Date'] = pd.to_datetime(store['Ship Date'])

In [62]:
store['Delivery(days)'] = store['Ship Date'] - store['Order Date']

In [63]:
store.head()

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88           0.50            2   
1   20847           High      0.01        2.84           0.93            3   
2   23086  Not Specified      0.03        6.68           6.15            3   
3   23087  Not Specified      0.01        5.68           3.60            3   
4   23088  Not Specified      0.00      205.99           2.50            3   

     Customer Name    Ship Mode Customer Segment Product Category  \
0  Janice Fletcher  Regular Air        Corporate  Office Supplies   
1    Bonnie Potter  Express Air        Corporate  Office Supplies   
2    Bonnie Potter  Express Air        Corporate  Office Supplies   
3    Bonnie Potter  Regular Air        Corporate  Office Supplies   
4    Bonnie Potter  Express Air        Corporate       Technology   

            Product Sub-Category Product Container  \
0                         Labels         Small Box   
1            Pens & Art Supplies          Wrap Bag   
2                          Paper         Small Box   
3  Scissors, Rulers and Trimmers        Small Pack   
4   Telephones and Communication         Small Box   

                                     Product Name  Product Base Margin  \
0                                        Avery 49                 0.36   
1  SANFORD Liquid Accent™ Tank-Style Highlighters                 0.54   
2                                      Xerox 1968                 0.37   
3        Acme® Preferred Stainless Steel Scissors                 0.56   
4                                             V70                 0.59   

    Region State or Province       City  Postal Code Order Date  Ship Date  \
0  Central          Illinois    Addison        60101 2012-05-28 2012-05-30   
1     West        Washington  Anacortes        98221 2010-07-07 2010-07-08   
2     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
3     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
4     West        Washington  Anacortes        98221 2011-07-27 2011-07-27   

     Profit  Quantity ordered new    Sales  Order ID Delivery(days)  
0    1.3200                     2     5.90     88525         2 days  
1    4.5600                     4    13.01     88522         1 days  
2  -47.6400                     7    49.92     88523         1 days  
3  -30.5100                     7    41.64     88523         1 days  
4  998.2023                     8  1446.67     88523         0 days

Q9 Based on Q8 find out for which order id we have shipment duration of more than 10 days.

In [68]:
store[store['Delivery(days)'] > '10 days']

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
643    18246       Critical      0.05       11.09           5.25          250   
1548   23091         Medium      0.04      100.98           7.18          595   
1549   23092         Medium      0.09        6.48           7.49          595   
1678   20703  Not Specified      0.07      130.98          30.00          637   
1679   20704  Not Specified      0.09       55.98           4.86          637   
1680   20705  Not Specified      0.08      115.99           4.23          637   
1697    2703  Not Specified      0.07      130.98          30.00          640   
1698    2704  Not Specified      0.09       55.98           4.86          640   
1699    2705  Not Specified      0.08      115.99           4.23          640   
2515   18060            Low      0.02        2.21           1.00          964   
5548   21603  Not Specified      0.02        5.58           1.99         2048   
5673   19005            Low      0.06       65.99           2.50         2088   
5859   24392            Low      0.05       29.89           1.99         2143   
5881   21257            Low      0.05        7.99           5.03         2153   
8607   26351         Medium      0.00        8.32           2.38         3119   
8609   26350         Medium      0.07      180.98          55.24         3119   
8610   26352         Medium      0.06        4.20           2.26         3119   
8973   20750            Low      0.06      125.99           8.08         3248   
8982    2748            Low      0.03       11.97           5.81         3251   
8983    2749            Low      0.08      195.99           4.20         3251   
8993    2750            Low      0.06      125.99           8.08         3251   
8996   20748            Low      0.03       11.97           5.81         3252   
8997   20749            Low      0.08      195.99           4.20         3252   

                Customer Name       Ship Mode Customer Segment  \
643   Brenda Nelson Blanchard     Regular Air        Corporate   
1548               Dana Desai     Regular Air   Small Business   
1549               Dana Desai     Express Air   Small Business   
1678       Christopher Bryant  Delivery Truck         Consumer   
1679       Christopher Bryant     Regular Air         Consumer   
1680       Christopher Bryant     Regular Air         Consumer   
1697               Neal Wolfe  Delivery Truck         Consumer   
1698               Neal Wolfe     Regular Air         Consumer   
1699               Neal Wolfe     Regular Air         Consumer   
2515          Virginia Rivera     Regular Air      Home Office   
5548          Ricky Dougherty     Regular Air         Consumer   
5673             Carrie Boyle     Regular Air        Corporate   
5859            Lester Sawyer     Regular Air      Home Office   
5881               Anna Glenn     Regular Air      Home Office   
8607              Jay Hubbard     Express Air        Corporate   
8609              Jay Hubbard  Delivery Truck        Corporate   
8610              Jay Hubbard     Regular Air        Corporate   
8973            Earl Donnelly     Express Air        Corporate   
8982             Peter Brooks     Regular Air        Corporate   
8983             Peter Brooks     Regular Air        Corporate   
8993             Peter Brooks     Express Air        Corporate   
8996           Milton Harrell     Regular Air        Corporate   
8997           Milton Harrell     Regular Air        Corporate   

     Product Category          Product Sub-Category Product Container  \
643   Office Supplies                     Envelopes         Small Box   
1548       Technology          Computer Peripherals         Small Box   
1549  Office Supplies                         Paper         Small Box   
1678        Furniture            Chairs & Chairmats        Jumbo Drum   
1679  Office Supplies                         Paper         Small Box   
1680       Technology  Telephones and 

In [69]:
store[store['Delivery(days)'] > '10 days']['Order ID']

643     87215
1548    86318
1549    86318
1678    87957
1679    87957
1680    87957
1697    19556
1698    19556
1699    19556
2515    86177
5548    88223
5673    88352
5859    87572
5881    91294
8607    86434
8609    86436
8610    86436
8973    87300
8982    19841
8983    19841
8993    19841
8996    87300
8997    87300
Name: Order ID, dtype: int64

Q10 Try to find out the list of returned order for which shipment duration was more then 15 days and find out that region manager as well.

In [71]:
store[store['Delivery(days)'] > '15 days'][['Order ID', 'Customer Segment']]

Order ID Customer Segment
643      87215        Corporate
1678     87957         Consumer
1680     87957         Consumer
1697     19556         Consumer
1699     19556         Consumer
2515     86177      Home Office
5673     88352        Corporate
5859     87572      Home Office
5881     91294      Home Office
8607     86434        Corporate
8609     86436        Corporate
8610     86436        Corporate
8973     87300        Corporate
8982     19841        Corporate
8983     19841        Corporate
8993     19841        Corporate
8996     87300        Corporate
8997     87300        Corporate

In [73]:
df = store[store['Delivery(days)'] > '15 days'][['Order ID', 'Customer Segment']]

In [74]:
df['Customer Segment'].value_counts()

Corporate      11
Consumer        4
Home Office     3
Name: Customer Segment, dtype: int64

Q11 Group by region and find out which region is more profitable.

In [75]:
store.groupby('Region')['Profit'].sum()

Region
Central    519825.567067
East       377566.186045
South      104201.192420
West       310849.453897
Name: Profit, dtype: float64

Central Region is the most profitable.

Q12 Try to find out overall in which country we are giving more discount.

In [76]:
store.head()

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88           0.50            2   
1   20847           High      0.01        2.84           0.93            3   
2   23086  Not Specified      0.03        6.68           6.15            3   
3   23087  Not Specified      0.01        5.68           3.60            3   
4   23088  Not Specified      0.00      205.99           2.50            3   

     Customer Name    Ship Mode Customer Segment Product Category  \
0  Janice Fletcher  Regular Air        Corporate  Office Supplies   
1    Bonnie Potter  Express Air        Corporate  Office Supplies   
2    Bonnie Potter  Express Air        Corporate  Office Supplies   
3    Bonnie Potter  Regular Air        Corporate  Office Supplies   
4    Bonnie Potter  Express Air        Corporate       Technology   

            Product Sub-Category Product Container  \
0                         Labels         Small Box   
1            Pens & Art Supplies          Wrap Bag   
2                          Paper         Small Box   
3  Scissors, Rulers and Trimmers        Small Pack   
4   Telephones and Communication         Small Box   

                                     Product Name  Product Base Margin  \
0                                        Avery 49                 0.36   
1  SANFORD Liquid Accent™ Tank-Style Highlighters                 0.54   
2                                      Xerox 1968                 0.37   
3        Acme® Preferred Stainless Steel Scissors                 0.56   
4                                             V70                 0.59   

    Region State or Province       City  Postal Code Order Date  Ship Date  \
0  Central          Illinois    Addison        60101 2012-05-28 2012-05-30   
1     West        Washington  Anacortes        98221 2010-07-07 2010-07-08   
2     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
3     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
4     West        Washington  Anacortes        98221 2011-07-27 2011-07-27   

     Profit  Quantity ordered new    Sales  Order ID Delivery(days)  
0    1.3200                     2     5.90     88525         2 days  
1    4.5600                     4    13.01     88522         1 days  
2  -47.6400                     7    49.92     88523         1 days  
3  -30.5100                     7    41.64     88523         1 days  
4  998.2023                     8  1446.67     88523         0 days

In [78]:
store.groupby('State or Province')['Discount'].mean()

State or Province
Alabama                 0.046960
Arizona                 0.051194
Arkansas                0.047154
California              0.051205
Colorado                0.045480
Connecticut             0.052195
Delaware                0.044000
District of Columbia    0.045588
Florida                 0.050536
Georgia                 0.045748
Idaho                   0.048860
Illinois                0.050873
Indiana                 0.050705
Iowa                    0.045833
Kansas                  0.051805
Kentucky                0.050964
Louisiana               0.054607
Maine                   0.052031
Maryland                0.049719
Massachusetts           0.050541
Michigan                0.046758
Minnesota               0.052385
Mississippi             0.049872
Missouri                0.050248
Montana                 0.048776
Nebraska                0.046623
Nevada                  0.041860
New Hampshire           0.048519
New Jersey              0.049322
New Mexico              0

In [79]:
df = pd.DataFrame(store.groupby('State or Province')['Discount'].mean())

In [84]:
df['Discount'] = df['Discount'] * 100

In [87]:
df

Discount  rank
State or Province                   
Alabama               4.696000  40.0
Arizona               5.119403  12.0
Arkansas              4.715447  39.0
California            5.120470  11.0
Colorado              4.548023  47.0
Connecticut           5.219512   7.0
Delaware              4.400000  48.0
District of Columbia  4.558824  46.0
Florida               5.053640  18.0
Georgia               4.574766  45.0
Idaho                 4.885965  31.0
Illinois              5.087329  14.0
Indiana               5.070539  16.0
Iowa                  4.583333  44.0
Kansas                5.180451   9.0
Kentucky              5.096386  13.0
Louisiana             5.460674   2.0
Maine                 5.203125   8.0
Maryland              4.971910  26.0
Massachusetts         5.054054  17.0
Michigan              4.675841  41.0
Minnesota             5.238494   5.0
Mississippi           4.987179  24.0
Missouri              5.024845  21.0
Montana               4.877551  32.0
Nebraska              4.662338  42.0
Nevada                4.186047  49.0
New Hampshire         4.851852  35.0
New Jersey            4.932203  29.0
New Mexico            4.857143  33.0
New York              5.027875  20.0
North Carolina        4.788845  36.0
North Dakota          5.029412  19.0
Ohio                  4.977273  25.0
Oklahoma              4.971154  27.0
Oregon                5.011905  23.0
Pennsylvania          4.915129  30.0
Rhode Island          5.450000   3.0
South Carolina        5.085714  15.0
South Dakota          4.642857  43.0
Tennessee             5.150602  10.0
Texas                 4.854489  34.0
Utah                  5.253425   4.0
Vermont               5.229508   6.0
Virginia              4.949495  28.0
Washington            4.733945  38.0
West Virginia         5.023256  22.0
Wisconsin             4.769231  37.0
Wyoming               5.476190   1.0

Q13 Give me list of unique postal codes.

In [90]:
store['Postal Code'].value_counts()

10177    54
90049    47
20016    37
60601    36
90045    33
         ..
20602     1
1757      1
6241      1
6488      1
2638      1
Name: Postal Code, Length: 1697, dtype: int64

In [91]:
store['Postal Code'].unique()

array([60101, 98221, 91776, ..., 61832, 62521, 26554])

Q14 Which customer segement is most profitalble find it out. 

In [92]:
store.groupby('Customer Segment')['Profit'].sum()

Customer Segment
Consumer          206559.625348
Corporate         505538.627783
Home Office       283869.553814
Small Business    316474.592482
Name: Profit, dtype: float64

Corporate Segment is the most profitable.

Q15 Try to find out the 10th most loss making product catagory.

In [94]:
store.groupby('Product Name')['Profit'].sum()

Product Name
"While you Were Out" Message Book, One Form per Page     708.820800
#10 Self-Seal White Envelopes                            223.197520
#10 White Business Envelopes,4 1/8 x 9 1/2              1164.001460
#10- 4 1/8" x 9 1/2" Recycled Envelopes                 -888.757800
#10- 4 1/8" x 9 1/2" Security-Tint Envelopes           -1288.400800
                                                           ...     
i500plus                                                 808.244640
i600                                                    -525.934816
iDEN i550                                               2624.076000
iDEN i95                                                -371.274704
iDENi80s                                                -471.229440
Name: Profit, Length: 1263, dtype: float64

In [95]:
df = pd.DataFrame(store.groupby('Product Name')['Profit'].sum())

In [98]:
df['Rank'] = df['Profit'].rank(method = 'max')

In [103]:
df

Profit    Rank
Product Name                                                           
"While you Were Out" Message Book, One Form per...   708.820800   817.0
#10 Self-Seal White Envelopes                        223.197520   684.0
#10 White Business Envelopes,4 1/8 x 9 1/2          1164.001460   912.0
#10- 4 1/8" x 9 1/2" Recycled Envelopes             -888.757800   180.0
#10- 4 1/8" x 9 1/2" Security-Tint Envelopes       -1288.400800   132.0
...                                                         ...     ...
i500plus                                             808.244640   838.0
i600                                                -525.934816   259.0
iDEN i550                                           2624.076000  1052.0
iDEN i95                                            -371.274704   306.0
iDENi80s                                            -471.229440   271.0

[1263 rows x 2 columns]

In [105]:
df[df['Rank'] == 10]

Profit  Rank
Product Name                                                       
Eldon ClusterMat Chair Mat with Cordless Antist... -9543.6525  10.0

Loss made by this company is 9543.6525

Q16 Try to find out 10 top  product with highest margins.

In [106]:
store.head()

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88           0.50            2   
1   20847           High      0.01        2.84           0.93            3   
2   23086  Not Specified      0.03        6.68           6.15            3   
3   23087  Not Specified      0.01        5.68           3.60            3   
4   23088  Not Specified      0.00      205.99           2.50            3   

     Customer Name    Ship Mode Customer Segment Product Category  \
0  Janice Fletcher  Regular Air        Corporate  Office Supplies   
1    Bonnie Potter  Express Air        Corporate  Office Supplies   
2    Bonnie Potter  Express Air        Corporate  Office Supplies   
3    Bonnie Potter  Regular Air        Corporate  Office Supplies   
4    Bonnie Potter  Express Air        Corporate       Technology   

            Product Sub-Category Product Container  \
0                         Labels         Small Box   
1            Pens & Art Supplies          Wrap Bag   
2                          Paper         Small Box   
3  Scissors, Rulers and Trimmers        Small Pack   
4   Telephones and Communication         Small Box   

                                     Product Name  Product Base Margin  \
0                                        Avery 49                 0.36   
1  SANFORD Liquid Accent™ Tank-Style Highlighters                 0.54   
2                                      Xerox 1968                 0.37   
3        Acme® Preferred Stainless Steel Scissors                 0.56   
4                                             V70                 0.59   

    Region State or Province       City  Postal Code Order Date  Ship Date  \
0  Central          Illinois    Addison        60101 2012-05-28 2012-05-30   
1     West        Washington  Anacortes        98221 2010-07-07 2010-07-08   
2     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
3     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
4     West        Washington  Anacortes        98221 2011-07-27 2011-07-27   

     Profit  Quantity ordered new    Sales  Order ID Delivery(days)  
0    1.3200                     2     5.90     88525         2 days  
1    4.5600                     4    13.01     88522         1 days  
2  -47.6400                     7    49.92     88523         1 days  
3  -30.5100                     7    41.64     88523         1 days  
4  998.2023                     8  1446.67     88523         0 days

In [107]:
store.groupby('Product Name')['Product Base Margin'].mean()

Product Name
"While you Were Out" Message Book, One Form per Page    0.35
#10 Self-Seal White Envelopes                           0.36
#10 White Business Envelopes,4 1/8 x 9 1/2              0.38
#10- 4 1/8" x 9 1/2" Recycled Envelopes                 0.38
#10- 4 1/8" x 9 1/2" Security-Tint Envelopes            0.36
                                                        ... 
i500plus                                                0.58
i600                                                    0.56
iDEN i550                                               0.60
iDEN i95                                                0.59
iDENi80s                                                0.59
Name: Product Base Margin, Length: 1263, dtype: float64

In [110]:
df = pd.DataFrame(store.groupby('Product Name')['Product Base Margin'].mean())

Converting margins into percentage

In [112]:
df['Product Base Margin'] = df['Product Base Margin']* 100

In [113]:
df['Rank'] = df['Product Base Margin'].rank(method = 'max', ascending = False)

In [116]:
df[df['Rank'] <= 10]

Product Base Margin  Rank
Product Name                                                                 
Accessory27                                                        85.0   4.0
Accessory4                                                         85.0   5.0
Belkin OmniView SE Rackmount Kit                                   85.0   4.0
Office Impressions Heavy Duty Welded Shelving &...                 85.0   4.0
Serrated Blade or Curved Handle Hand Letter Ope...                 84.0   8.0
Sterling Rubber Bands by Alliance                                  85.0   4.0
Tennsco Commercial Shelving                                        84.0   8.0
Tennsco Stur-D-Stor Boltless Shelving, 5 Shelve...                 84.0   8.0